File: code/day3/go/duplication_in_array_not_edit.go
package main

import "fmt"

func countRange(numbers []int, length int, start, end int) int {
	count := 0
	for _, num := range numbers {
		if (num >= start && num <= end) {
			count++
		}
	}
	return count
}

func findDuplicationNumberInArray(numbers []int, length int) int {
	if (length <= 0) {
		return -1
	}

	for i:=0; i < length; i++ {
		if (numbers[i] < 0 || numbers[i] > length -1) {
			return -1
		}
	}

	start := 1
	end := length - 1
	for start <= end {
		middle := ((end - start) >> 1) + start
		count := countRange(numbers, length, start, middle)
		if end == start {
			if count > 1 {
				return start
			} else {
				break
			}
		}

		if count > (middle- start +1) {
			end = middle
		} else {
			start = middle + 1
		}
	}
	return -1
}

func main() {
	var numbers = []int{2, 3, 1, 6, 5, 2, 3}
	fmt.Println("duplication number: ", findDuplicationNumberInArray(numbers, len(numbers)))
	var no_duplication_numbers = []int{2, 3, 1, 7, 5, 6, 4}
	fmt.Println("duplication number: ", findDuplicationNumberInArray(no_duplication_numbers, len(no_duplication_numbers)))
}

File: code/c/day3/duplication_in_array_not_edit.c
#include<stdlib.h>
#include<stdio.h>

// 计算区间内数字的多少
int count_range(int* numbers, int length, int start, int end)
{
    if(numbers == NULL)
        return 0;

    int count = 0;
    for(int i=0; i<length; i++)
    {
        if(numbers[i] >= start && numbers[i] <= end)
        {
            ++count;
        }
    }
    return count;
}

int find_duplication_number_in_array(int numbers[], int length)
{
    if (length <= 0)
        return -1;
    
    for(int i=0; i < length; ++i)
    {
        if(numbers[i] < 0 || numbers[i] > length - 1)
            return -1;
    }

    int start = 1;
    int end = length - 1;
    while(end >= start)
    {
        int middle = ((end - start) >> 1) + start;
        int count = count_range(numbers, length, start, middle);
        if(end == start)  // 区间只剩下一个数字
        {
            if(count > 1)
                return start;
            else
                break;
        }
        if(count > (middle - start + 1))
            end = middle;
        else
            start = middle + 1;
    }
    return -1;
}

void test_0()
{
    int numbers[] = {2, 3, 1, 6, 5, 2, 3};
    printf("duplication number: %d\n", find_duplication_number_in_array(numbers, sizeof(numbers)/sizeof(numbers[0])));
}

void test_1()
{
    int numbers[] = {2, 3, 1, 7, 5, 6, 4};
    printf("duplication number: %d\n", find_duplication_number_in_array(numbers, sizeof(numbers)/sizeof(numbers[0])));
}

void test_2()
{
    int numbers[] = {1, 5, 6};
    printf("duplication number: %d\n", find_duplication_number_in_array(numbers, sizeof(numbers)/sizeof(numbers[0])));
}

void run_all_test()
{
    test_0();
    test_1();
    test_2();
}

int main(int argc, char* argv[])
{
  run_all_test();
  return EXIT_SUCCESS;
}